pkg/ui/workspaces/cluster-ui/src/api/databases/tableMetaUpdateJobApi.ts

```diff
@@ -0,0 +1,141 @@
+};
+
+type TableMetaUpdateJobInfo = {
+  currentStatus: TableMetaUpdateJobResponse["current_status"];
+  progress: number;
+  lastStartTime: moment.Moment;
+  lastCompletedTime: moment.Moment;
+  lastUpdatedTime: moment.Moment;
+  dataValidDuration: moment.Duration;
+  automaticUpdatesEnabled: boolean;
+};
+
+const getTableMetaUpdateJobInfo = async () => {
+  return fetchDataJSON<TableMetaUpdateJobResponse, null>(TABLE_META_JOB_API);
+};
+
+// useTableMetaUpdateJob is a hook that fetches the current status of the table
+// metadata update job and returns the status and other relevant information.
+// The hook polls the API every 3 seconds if the job status is parsed as running.
+export const useTableMetaUpdateJob = () => {
+  const { data, isLoading, mutate } = useSWR(
+    "tableMetaUpdateJob",
+    getTableMetaUpdateJobInfo,
+    {
+      focusThrottleInterval: 10000,
+      refreshInterval: (latest: TableMetaUpdateJobResponse) => {
+        return latest?.current_status === TableMetadataJobStatus.RUNNING
+          ? 3000
+          : 0;
+      },
+      dedupingInterval: 3000,
+    },
+  );
+
+  const formattedResp: TableMetaUpdateJobInfo = useMemo(() => {
+    if (!data) {
+      return null;
+    }
+    return {
+      currentStatus: data.current_status,
+      progress: data.progress,
+      lastStartTime: data.last_start_time
+        ? moment.utc(data.last_start_time)
+        : null,
+      lastCompletedTime: data.last_completed_time
+        ? moment.utc(data.last_completed_time)
+        : null,
+      lastUpdatedTime: data.last_updated_time
+        ? moment.utc(data.last_updated_time)
+        : null,
+      dataValidDuration: moment.duration(
+        data.data_valid_duration * 1e-6,
+        "millisecond",
+      ),
+      automaticUpdatesEnabled: data.automatic_updates_enabled,
+    };
+  }, [data]);
+
+  const dataValidDurationMs =
+    formattedResp?.dataValidDuration?.asMilliseconds();
+  const lastCompletedTimeUnixMs =
+    formattedResp?.lastCompletedTime?.unix() * 1000;
+  useEffect(() => {
+    if (isLoading) {
+      return;
+    }
+
+    // Trigger a refresh a little after when the data is scheduled to
+    // expire. We add a 3 second slack in case an update is triggered
+    // around then.
+    const msSinceLastCompletedWithSlack =
+      Date.now() - lastCompletedTimeUnixMs + 3000;
+    const delay = Math.max(
+      0,
+      dataValidDurationMs - msSinceLastCompletedWithSlack,
+    );
+    const nextUpdated = setTimeout(() => {
+      return mutate();
+    }, delay);
+
+    return () => clearTimeout(nextUpdated);
+  }, [dataValidDurationMs, isLoading, lastCompletedTimeUnixMs, mutate]);
+
+  return {
+    jobStatus: formattedResp,
+    refreshJobStatus: mutate,
+    isLoading,
+  };
+};
+
+type TriggerTableMetaUpdateJobRequest = {
+  onlyIfStale?: boolean;
+};
+type TriggerTableMetaUpdateJobResponse = {
+  job_triggered: boolean;
+  message: string;
+};
+
+export const triggerUpdateTableMetaJobApi = async (
```

Comment:
Why do we need to trigger it from the UI automatically? Why can't the server just do it anytime it's stale?

Reply:
Here's what I see as the two behavior possibilities:

* When button is clicked: refresh always happens
* All other cases: server decides if a refresh is necessary based on staleness

Why is there a separate flag needed to say "use the staleness trigger"?

Reply:
I think I'm still not following - the flag is part of the request to the job trigger api which tells the server which case we fall into, otherwise I'm not sure how we'd distinguish between a 'force' request and one that should adhere to the staleness threshold.